boofuzz/primitives/string.py

```diff
@@ -223,24 +222,43 @@ def _yield_long_strings(self, sequences):
         for sequence in sequences:
             for size in self._long_string_lengths:
                 for delta in self._long_string_deltas:
-                    yield sequence * (size + delta)
-
-        for size in self._long_string_lengths:
-            s = "D" * size
-            for loc in self.random_indices[size]:
-                s = s[:loc] + "\x00" + s[loc:]
-                yield s
+                    if self.max_len is None or size + delta <= self.max_len:
+                        data = sequence * math.ceil((size + delta) / len(sequence))
+                        yield data[: size + delta]
 
-        for sequence in sequences:
             for size in self._extra_long_string_lengths:
-                yield sequence * size
+                if self.max_len is None or size <= self.max_len:
+                    data = sequence * math.ceil(size / len(sequence))
+                    yield data[:size]
+
+            if self.max_len is not None and (
+                self.max_len < min(self._long_string_lengths[0], self._extra_long_string_lengths[0])
+                or self.max_len
+                > max(self._long_string_lengths[-1] + self._long_string_deltas[-1], self._extra_long_string_lengths[-1])
+            ):
+                data = sequence * math.ceil(self.max_len / len(sequence))
+                yield data[: self.max_len]
+
+        for size in self._long_string_lengths:
+            if self.max_len is None or size <= self.max_len + 1:
+                s = "D" * size
+                for loc in self.random_indices[size]:
+                    s = s[:loc] + "\x00" + s[loc:]  # Do we really have to insert a char or can we replace it?
+                    yield s
 
     def _yield_variable_mutations(self, default_value):
         for length in self._variable_mutation_multipliers:
-            yield default_value * length
-        # doesn't make sense while we're yielding strings:
-        # for length in self._variable_mutation_multipliers:
-        #     yield default_value * length + b"\xfe"
+            value = default_value * length
+            if value not in self._fuzz_library:
+                yield value
+                if self.max_len is not None and self.max_len <= len(value):
```

Comment:
The purpose of the max_len checks is to prevent the generation of data which is too long and would be dropped later.
I found the idea of fist generating all possible data and later filtering it really inefficient. Considering that we generate up to 1MB large strings I think this could also have a noticeable performance impact.

But I agree that in some cases it's kinda hard to wrap your head around all these checks and conditions.

So in conclusion, the max_len checks improve the performance at the cost of readability. Not sure if we should remove them.

Reply:
The last_val duplicate check is intended to filter duplicates in the fuzz_library which are a result of truncation.
It's main purpose is not to filter the data which we generate by multiplying sequences and sizes

Reply:
In `_yield_long_strings` I didn't want to yield data that would be truncated later. That's why I placed those max_len checks in there and added a break to the for loops to stop unnecessary iterations.

Reply:
Works for me. Just to clarify, does this mean that for `_yield_long_strings()`, you won't get a long string exactly equal to the max_len?

Reply:
@SR4ven Nevermind, I see now you changed that elsewhere. Good to go!

Reply:
It will yield all lengths specified in `_long_string_lengths` and `_extra_long_string_lengths` plus one string exactly max_len long from here:
https://github.com/jtpereyda/boofuzz/blob/c338bf69a500018658d528db3432ecd36d4a1217/boofuzz/primitives/string.py#L240-L242